docs/design/1443-dynamic-pipelines.md

```diff
@@ -0,0 +1,252 @@
+  // E.g. `target["__address__"]`.
+  var = "target"
+
+  // A sub-pipeline consisting of components which process each target.
+  ...
+}
+```
+
+<details>
+  <summary>Example</summary>
+
+```alloy
+discovery.file "default" {
+  files = ["/Users/batman/Desktop/redis_addresses.yaml"]
+}
+
+// Every component defined in the "foreach" block will be instantiated for each item in the collection. 
+// The instantiated components will be scoped using the name of the foreach block and the index of the
+// item in the collection. For example: /foreach.redis/0/prometheus.exporter.redis.default
+foreach "redis" {
+  collection = discovery.file.default.targets
+  // Here, "target" is a variable whose value is the current item in the collection.
+  var = "target"
+
+  prometheus.exporter.redis "default" {
+    redis_addr = target["__address__"] // we can also do the necessary rewrites before this.
+  }
+
+  discovery.relabel "default" {
+    targets = prometheus.exporter.redis.default.targets
+    // Add a label which comes from the discovery component.
+    rule {
+      target_label = "filepath"
+      // __meta_filepath comes from discovery.file
+      replacement  = target["__meta_filepath"]
+    }
+  }
+
+  prometheus.scrape "default" {
+    targets = discovery.relabel.default.targets
+    forward_to = prometheus.remote_write.mimir.receiver
+  }
+}
+
+prometheus.remote_write "mimir" {
+  endpoint {
+    url = "https://prometheus-prod-05-gb-south-0.grafana.net/api/prom/push"
+    basic_auth {
+      username = ""
+      password = ""
+    }
+  }
+}
+```
+
+</details>
+
+Pros:
+* The `foreach` name is consistent with other programming languages.
+
+Cons:
+* It looks less like a component than a `declare.dynamic` block.
+  In order to instantiate multiple `foreach` blocks with similar config, you'd have to wrap them in a `declare` block.
+
+## Proposal 2: A declare.dynamic block
+
+A new `declare.dynamic` block would create a custom component which starts several sub-pipelines internally.
+Users can use `argument` and `export` blocks, just like in a normal `declare` block.
+
+```alloy
+declare.dynamic "ex1" {
+  argument "input_targets" {
+    optional = false
+    comment = "We will create a sub-pipeline for each target in input_targets."
+  }
+
+  argument "output_metrics" {
+    optional = false
+    comment = "All the metrics gathered from all pipelines."
+  }
+
+  // A sub-pipeline consisting of components which process each target.
+  ...
+}
+
+declare.dynamic.ex1 "default" {
+  input_targets = discovery.file.default.targets
+  output_metrics = [prometheus.remote_write.mimir.receiver]
+}
+```
+
+<details>
+  <summary>Example</summary>
+
+```alloy
+// declare.dynamic "maps" each target to a sub-pipeline.
+// Each sub-pipeline has 1 exporter, 1 relabel, and 1 scraper.
+// Internally, maybe one way this can be done via serializing the pipeline to a string and then importing it as a module?
```

Comment:
that's an interesting idea. One way to do this could be similar to the import node. The declare.dynamic instance would manage a list of custom component nodes that it can run using the runner package. These custom component nodes would not be part of the initial graph and would take the pipeline template as AST body to run.